Light state

zsfelfoldi opened the pull request:
This PR implements ODR capable state & trie structures for the light client.

robotally commented:
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Tue Dec 1 03:54:44 UTC 2015_

zsfelfoldi commented:
@obscuren fixed the issues you mentioned and rebased on develop

fjl commented on light/state_object.go:
```diff
@@ -0,0 +1,233 @@
+	// Temporarily initialisation code
+	initCode Code
+	// Cached storage (flushed when updated)
+	storage Storage
```
We need to find a way to share some of the logic with package state.
Most of the code (decoding, accessors, etc.) is simply copied from package state.

Maybe one way to do it would be to simplify `state.StateObject` so it doesn't contain a DB.
In that case, you could simply embed it here and get all the boring methods for free.

fjl replied:
On second thought, we could also address the duplication later when the protocol is in.

zsfelfoldi replied:
The highest priority is to not mess up existing stuff so I'd definitely wait with that until we have a functional light client. After everything is built up (including the VM calls) we can make better decisions about what code can be shared and how that should be done.

fjl commented on light/trie.go:
```diff
@@ -0,0 +1,126 @@
+		nibbles := append(mn.KeyPrefix, mn.KeySuffix...)
+		var key []byte
+		if len(nibbles) > 0 {
+			key = compactHexEncode(nibbles)
```
This is weird. If we need the full encoded key, we might just as well put it in the error when it is created.

fjl commented on trie/proof.go:
```diff
@@ -12,6 +12,8 @@ import (
 	"github.com/ethereum/go-ethereum/rlp"
 )
 
+type MerkleProof []rlp.RawValue
```
I think this type definition doesn't add any value. You can always pass merkle proofs as `[]rlp.RawValue`.
Type definitions make it possible to attach methods to a type, but `MerkleProof` doesn't have any.

zsfelfoldi replied:
changed back to []rlp.RawValue

fjl commented on trie/proof.go:
```diff
@@ -93,7 +94,12 @@ func VerifyProof(rootHash common.Hash, key []byte, proof []rlp.RawValue) (value
+				return nil, fmt.Errorf("key mismatch at proof node %d", i)
+			} else {
+				// The trie doesn't contain the key.
+				return nil, nil
```
Why is this useful (just curious)?

fjl replied:
Looking at the changes to `Prove` again, I see that you changed it as well to return a non-nil proof if the key does not exist.

zsfelfoldi replied:
A missing key is a perfectly valid zero-value entry so we need to be able to prove it. Updated the comments of Prove.

zsfelfoldi commented:
@fjl the issues you mentioned are fixed now

fjl commented:
:+1:

zelig commented on light/odr.go:
```diff
@@ -0,0 +1,106 @@
+}
+
+// OdrRequest is an interface for retrieval requests
+type OdrRequest interface {
```
if you wrap the request in a struct, you can DRY up all the GetCtx redundancies.

```go
type OdrBackend interface {
   Database() ethdb.Database
  Retrieve(req *OdrRequestWithCtx)
}
type OdrRequestWithCtx struct {
 ctx   context.Context
 OdrRequest
}

type OdrRequest interface {
  StoreResult(db ethdb.Database)
}
// GetCtx not needs to be defined exactly once not 5 times
func (req *OdrRequestWithCtx) GetCtx() context.Context { return req.ctx }


// this now does not need the ctx field and the GetCtx method to be defined
type TrieRequest struct {
	root  common.Hash
	key   []byte
	proof []rlp.RawValue
}
```

```

But the question is, do you really need the ctx embedded in the request? cant you always just pass it to Retrieve if it was OdrBackend#Retrieve(req OdrRequest, context.Context)
and eliminate the extra wrapping, GetCtx and ctx fields altogether.

zsfelfoldi replied:
You're right, that was a remnant of an old design. I don't want it wrapped in a struct but contexts shouldn't be wrapped anyways so they are now directly passed to Retrieve.

zelig commented:
:+1: great stuff

obscuren commented:
@fjl renamed a few methods in https://github.com/ethereum/go-ethereum/pull/2064 such as `NewStateObjectFromBytes` to `DecodeObject`. Please change this PR so it matching the other.

Additionally you need to document the exported methods :-)

obscuren commented:
:-1: (just so it shown up for @robotally and doesn't get accidentally merged)

fjl commented:
@zsfelfoldi I'll update #2064 tonight so it is mergeable

zsfelfoldi commented:
@obscuren I updated the entire stateobject.go with all the rlp changes and added missing comments to trie.go. I did not yet comment all the state and stateobject functions because there's a shitload of them and they're not commented in the original version either :) I'll do those too of course but maybe we should do that in the two versions together in a separate PR. I can do it after both this PR and https://github.com/ethereum/go-ethereum/pull/2064 are merged. Or alternatively, if @fjl comments these functions in his PR, I'll copy those into mine, but I'm not saying that he should do it :)

obscuren commented:
I don't see why we should neglect on commenting because someone else has a year ago (me). Since this PR adds the methods we should add the documentation in this PR.

zsfelfoldi commented:
@obscuren ok, I'll do it in this PR today.

zsfelfoldi commented:
@obscuren done
